File: Library/Classes/Core/Categories/UIKit/UILabel+CSExtension.m
//
// Created by Rene Dohan on 02/12/14.
// Copyright (c) 2014 creative_studio. All rights reserved.
//

#import "UILabel+CSExtension.h"
#import "CSLang.h"
#import "UIView+CSDimension.h"
#import "UIView+CSLayoutGetters.h"
#import "NSString+CSExtension.h"
#import "UIView+CSExtension.h"

@implementation UILabel (CSExtension)

- (instancetype)construct {
    super.construct;
    self.lineBreakMode = NSLineBreakByTruncatingTail;
    return self;
}

- (instancetype)setFontSize :(CGFloat)size {
    self.font = [self.font fontWithSize :size];
    return self;
}

- (instancetype)fontSize :(CGFloat)size {
    self.font = [self.font fontWithSize :size];
    return self;
}

- (void)setFontStyle :(UIFontTextStyle)style {
    self.font = [UIFont preferredFontForTextStyle :style];
}

- (instancetype)fontStyle :(UIFontTextStyle)style {
    self.font = [UIFont preferredFontForTextStyle :style];
    return self;
}

- (instancetype)font :(UIFont *)font {
    self.font = font;
    return self;
}

- (instancetype)textColor :(UIColor *)textColor {
    self.textColor = textColor;
    return self;
}

- (instancetype)setHTMLFromString :(NSString *)string {
    string = [string add :stringf(
                  @"<style>body{font-family: '%@'; font-size:%fpx;}</style>", self.font.fontName, self.font.pointSize)];
    self.attributedText = [NSAttributedString.alloc initWithData :[string dataUsingEncoding :NSUnicodeStringEncoding]
                                                         options :@{ NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType,
                                                                     NSCharacterEncodingDocumentAttribute: @(NSUTF8StringEncoding) }
                                              documentAttributes :nil error :nil];
    return self;
}

- (instancetype)hideIfEmpty {
    [self setVisible :[NSString set :self.text]];
    return self;
}

- (instancetype)sizeHeightToLines :(int)numberOfLines {
//    NSAssert(self.width > 0, @"Width has to be set to calculate height");
    let currentWidth = self.width;
    let currentText = self.text;
    self.text = @"cjksjkljaskljfklsaj fjas klfjaslk jfklaj fklaj fkljs aklfj klasj"
        " fljsahflasljh sdiaf uiau fiahfiohe iof aeuhfkuaedfiuaehfueahkufheuafuaehfoiuyeaoif "
        "aeil fklaehjlfhaekjfhgkaegfjgeauklfeakuhfkluaehkfheaklufhkljaehfk hleauk "
        "fhlkuaehfjkaekfgeakfgkalehfkjahekjlfhkjelhqfkheukfglkgfalkjgfkagefjklgaekfgeajk"
        " fljsahflasljh sdiaf uiau fiahfiohe iof aeuhfkuaedfiuaehfueahkufheuafuaehfoiuyeaoif "
        "aeil fklaehjlfhaekjfhgkaegfjgeauklfeakuhfkluaehkfheaklufhkljaehfk hleauk "
        "fhlkuaehfjkaekfgeakfgkalehfkjahekjlfhkjelhqfkheukfglkgfalkjgfkagefjklgaekfgeajk"
        " fljsahflasljh sdiaf uiau fiahfiohe iof aeuhfkuaedfiuaehfueahkufheuafuaehfoiuyeaoif "
        "aeil fklaehjlfhaekjfhgkaegfjgeauklfeakuhfkluaehkfheaklufhkljaehfk hleauk "
        "fhlkuaehfjkaekfgeakfgkalehfkjahekjlfhkjelhqfkheukfglkgfalkjgfkagefjklgaekfgeajk"
        " fljsahflasljh sdiaf uiau fiahfiohe iof aeuhfkuaedfiuaehfueahkufheuafuaehfoiuyeaoif "
        "aeil fklaehjlfhaekjfhgkaegfjgeauklfeakuhfkluaehkfheaklufhkljaehfk hleauk "
        "fhlkuaehfjkaekfgeakfgkalehfkjahekjlfhkjelhqfkheukfglkgfalkjgfkagefjklgaekfgeajk";
    self.numberOfLines = numberOfLines;
    [self sizeToFit];
    self.text = currentText;
    self.width = currentWidth;
    return self;
}

- (instancetype)sizeFit :(NSString *)value {
    self.numberOfLines = 0;
    let current = self.text;
    return [[self text :value].sizeFit text :current];
}

- (instancetype)text :(NSString *)string {
    self.text = string;
    return self;
}

- (instancetype)textAlign :(enum NSTextAlignment)alignment {
    self.textAlignment = alignment;
    return self;
}

@end

File: Renetik/Renetik/Classes/Core/Extensions/UIKit/UILabel+CSExtension.swift
//
//  UILabel+CSExtension.swift
//  Renetik
//
//  Created by Rene Dohan on 3/6/19.
//

import UIKit
import RenetikObjc

public extension UILabel {

    class func construct(_ text: String) -> Self { construct().text(text) }

    override func construct() -> Self {
        super.construct(); text(lines: 0).text(break: .byTruncatingTail); return self
    }

    @discardableResult
    func text(color: UIColor) -> Self { self.textColor = textColor; return self }

    @discardableResult
    func text(align: NSTextAlignment) -> Self { textAlignment = align; return self }

    @discardableResult
    func text(break mode: NSLineBreakMode) -> Self { lineBreakMode = mode; return self }

    @discardableResult
    func text(lines: Int) -> Self { numberOfLines = lines; return self }

    @discardableResult
    override func heightToFit() -> Self { text(lines: 0); super.heightToFit(); return self }

    @discardableResult
    func resizeToFit(_ value: String) -> Self {
        text(lines: 0); let current = text; return text(value).resizeToFit().text(current ?? "")
    }

    @discardableResult
    func heightToFit(_ value: String) -> Self {
        text(lines: 0)
        let current = text; text = value; height = heightThatFits(); text = current
        return self
    }

    @discardableResult
    func heightToFit(lines numberOfLines: Int) -> Self {
        let currentWidth = width; let currentText = text; var linesText = "line"
        for i in 0..<numberOfLines - 1 { linesText += "\n line" }
        text(linesText).resizeToFit().text(currentText).width(currentWidth)
        return self
    }
}

extension UILabel: CSHasTextProtocol {
    public func text() -> String? { text }

    public func text(_ text: String?) { self.text = text }
}

extension UILabel: CSHasFontProtocol {
    public func font() -> UIFont? { font }

    public func font(_ font: UIFont?) { self.font = font }
}
